pkg/server/config/config.go

```diff
@@ -1,130 +1,83 @@
 
-	// k8s
-	KubeHost = LoadEnvVar(EnvKubeHost, "", true)
-	KubeConfig = LoadEnvVar(EnvKubeConfig, "", true)
+	// Size represents the capacity of the pvc, unit supports 'Gi' or 'Mi'
+	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#capacity
+	Size string `json:"size"`
+}
 
-	// log
-	LogLevel = LoadEnvVar(EnvLogLevel, DefaultLogLevel, true)
+// LoggingConfig configures logging
+type LoggingConfig struct {
+	Level string `json:"level"`
 }
 
-// GetStringEnvWithDefault retrieves the value of the environment variable named
-// by the key. If the variable is present in the environment the
-// value is returned. Otherwise the default is returned.
-func GetStringEnvWithDefault(name, defaultValue string) string {
-	v, ok := os.LookupEnv(name)
+// Config is Workflow Controller config instance
+var Config CycloneServerConfig
+
+// LoadConfig loads configuration from ConfigMap
+func LoadConfig(cm *core_v1.ConfigMap) error {
+	data, ok := cm.Data[ConfigFileKey]
 	if !ok {
-		return defaultValue
+		return fmt.Errorf("ConfigMap '%s' doesn't have data key '%s'", cm.Name, ConfigFileKey)
 	}
-	return v
-}
-
-// LoadEnvVar retrieves the value of 'name' frome environment.
-// If canBeEmpty is false and the value of the 'name' is empty, os will exit.
-func LoadEnvVar(name string, defaultValue string, canBeEmpty bool) string {
-	value := GetStringEnvWithDefault(name, defaultValue)
-	if len(value) == 0 && !canBeEmpty {
-		log.Fatalf(envVarCannotEmptyFormat, name)
-		os.Exit(-1)
+	err := json.Unmarshal([]byte(data), &Config)
+	if err != nil {
+		log.Warningf("Unmarshal config data %v error: %v", data, err)
+		return err
 	}
-	return value
-}
 
-// LoadFloatEnvVar ...
-func LoadFloatEnvVar(name string, defaultValue float64, canBeEmpty bool) float64 {
-	val := LoadEnvVar(name, "", canBeEmpty)
-	if len(val) == 0 {
-		return defaultValue
+	if !validate(&Config) {
+		return fmt.Errorf("validate config failed")
 	}
 
-	floatVal, err := strconv.ParseFloat(val, 64)
-	if err != nil {
-		log.Fatalf("The value '%s' of environment variable '%s' is not a valid float64: %v\n", val, name, err)
-		os.Exit(-1)
-	}
-	return floatVal
+	log.Infof("cyclone server config: %v", Config)
+	return nil
 }
 
-// LoadIntEnvVar ...
-func LoadIntEnvVar(name string, defaultValue int, canBeEmpty bool) int {
-	val := LoadEnvVar(name, "", canBeEmpty)
-	if len(val) == 0 {
-		return defaultValue
+// validate validates some required configurations.
+func validate(config *CycloneServerConfig) bool {
+	if config.CycloneServerHost == "" {
+		log.Warningf("CycloneServerHost not configured, cyclone server can not listen API requests.")
 	}
 
-	intVal, err := strconv.ParseInt(val, 10, 32)
-	if err != nil {
-		log.Fatalf("The value '%s' of environment variable '%s' is not a valid integer: %v\n", val, name, err)
-		os.Exit(-1)
+	if config.CycloneServerPort == 0 {
+		log.Warningf("CycloneServerPort not configured, cyclone server can not listen API requests.")
```

Comment:
If `cyclone server can not listen API requests`, should return `false` in my understand. And I think here we should fallback to default value. The same to server host, these two config items should be optional for users.

Reply:
OK